storage-provider/server/src/storage.rs

```diff
@@ -186,91 +192,94 @@ async fn upload(
-        };
-
-        let field_reader = StreamReader::new(field.map_err(std::io::Error::other));
-        stream_contents_to_car(state.car_piece_storage_dir.clone().as_ref(), field_reader)
+        // For multipart/form-data, we stream the field contents.
+        let state_clone = state.clone();
+        let stream: BoxedStream = Box::pin(async_stream::try_stream! {
+            let mut multipart = Multipart::from_request(request, &state_clone)
+                .await
+                .map_err(|err| io::Error::new(io::ErrorKind::Other, err.to_string()))?;
+            // Get the next field.
+            let mut field = multipart
+                .next_field()
+                .await
+                .map_err(|err| io::Error::new(io::ErrorKind::Other, err.to_string()))?
+                .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "empty request"))?;
+            // Yield each chunk as it becomes available.
+            while let Ok(Some(chunk)) = field.chunk().await {
+                yield chunk;
+            }
+        });
+        let field_reader = StreamReader::new(stream);
+        stream_contents_to_car(state.car_piece_storage_dir.as_ref(), field_reader)
             .await
             .map_err(|err| {
                 tracing::error!(%err, "failed to store file into CAR archive");
                 (StatusCode::INTERNAL_SERVER_ERROR, err.to_string())
             })?
     } else {
-        // Read the request body into a CAR archive
+        // For direct uploads, convert the request body into a stream.
         let body_reader = StreamReader::new(
             request
                 .into_body()
                 .into_data_stream()
                 .map_err(|err| io::Error::new(io::ErrorKind::Other, err)),
         );
-        stream_contents_to_car(state.car_piece_storage_dir.clone().as_ref(), body_reader)
+        stream_contents_to_car(state.car_piece_storage_dir.as_ref(), body_reader)
             .await
             .map_err(|err| {
                 tracing::error!(%err, "failed to store file into CAR archive");
                 (StatusCode::INTERNAL_SERVER_ERROR, err.to_string())
             })?
     };
-    tracing::debug!("generated cid: {file_cid}");
 
-    // NOTE(@jmg-duarte,03/10/2024): Maybe we should just register the file in RocksDB and keep a
-    // background process that vacuums the disk as necessary to simplify error handling here
+    tracing::debug!("generated cid: {file_cid}");
 
+    // Open the CAR file to check its size.
     let (_, file_path) = content_path(&state.car_piece_storage_dir, file_cid);
     let file = File::open(&file_path).await.map_err(|err| {
         tracing::error!(%err, path = %file_path.display(), "failed to open file");
         (StatusCode::INTERNAL_SERVER_ERROR, err.to_string())
     })?;
     let file_size = file
         .metadata()
-        .map_ok(|metadata| metadata.len())
         .await
+        .map(|m| m.len())
```

Comment:
1. Why are you moving this *past* the await point?
2. I get why you're shortening the variable name, but the full name keeps context. Revert